mipengine/controller/cleaner.py

```diff
@@ -78,41 +79,35 @@ def __init__(self):
         self._keep_cleaning_up = True
         self._cleanup_loop_thread = None
 
+    def cleanup_context_id(self, context_id, node_ids):
+        self._logger.debug(f"Cleaning up file with {context_id=}")
+        failed_node_ids = self._exec_context_id_cleanup(context_id, node_ids)
+
+        if failed_node_ids:
+            self._logger.debug(
+                f"'Altering' file with {context_id=}, removing node ids "
+                f"which succeeded cleanup:{set(node_ids) - set(failed_node_ids)=}, keeping only failed node ids: {failed_node_ids=}"
+            )
+            self._logger.debug(f"Deleting file with {context_id=}")
+            self._cleanup_files_processor.delete_file_by_context_id(context_id)
+            self._logger.debug(f"Re-creating file with {context_id=}")
+            self.add_contextid_for_cleanup(
```

Comment:
This is not a problem of this PR but if @apmariglis agrees this needs to be fixed.

If the CONTROLLER crashes in the middle of deleting and then creating again the context id file, there will be no information about the context id that needs to be cleaned up.

Reply:
yes what you mention is probably correct, but I think what is more concerning with this implementation is that the `CleanupFilesProcessor` class is not thread safe, which was not an issue before as it was only called by the thread created in the `Cleaner::start` method, but now it is also called from the thread where the `Contorller::exec_algorithm` method runs, which is a different thread. So I m not sure what will happen when in the middle of a call to `add_contextid_for_cleanup` method, which in turn calls `create_file_from_cleanup_entry`, another call to `add_contextid_for_cleanup` takes place. Also there are no such test but it does not seem very robust.

Reply:
Even with multiple calls for `add_contextid_for_cleanup` they will be for different context id so they will touch different files.
Is this the problem you had in mind?